docs/website/docs/developers/debugging/gpu-approaches-tips.md

```diff
@@ -0,0 +1,163 @@
+  GPU backend in LLVM proper or GPU driver compiler for SPIR-V.
+* For runtime, we have fewer layers--IREE HAL drivers, and GPU driver.
+
+Any of the above components/layers can have bugs.
+It's important to reduce the potential surface area to make the problem more
+tractable.
+
+Once we have a more isolated case, the general methodology to pinpoint the
+exact culprit is to **1) collect and inspect the symptoms, 2) form hypothesis
+and run experiments to prove/refute the hypothesis, and 3) iterate**.
+
+### .. with shortcuts
+
+The above procedure is for facing a large problem with no clue, for example,
+when bringing up a new model end-to-end via IREE.
+
+Though most of the time, we can leverage existing facilities to avoid going down
+the full top-down hiearchical debugging procedure.
+For example, for regression happening on an existing model, CI or `git bitsect`
+might tell us directly the culprit commit.
+
+### .. using tools
+
+For issues with strong signals like crashing, it's also easier to pinpoint the
+exact culprit with dedicated tools--we can leverage various
+[sanitizers](./sanitizers.md) or debuggers.
+
+## Isolating the problematic component
+
+If we are facing a large problem without a clear clue, we need to isolate the
+problematic compiler or runtime layer first, typically by comparing with a
+working solution:
+
+* `[correctness/performance]` Sanitize the environment first--did you recently
+  updated the GPU SDK or driver?
+  Are others able to reproduce the issue?
+  If not what SDK / driver versions they are using?
+  Is your machine drawing enough power when benchmarking?
+  Is your machine connected with a mointor (for Vulkan)?
+  How long since you last rebooted your machine? 👻
+  Asking these questions and making sure the environment is proper can save
+  you hours of debugging sometimes.
+* `[correctness/performance]` We have multiple GPU targets/drivers in
+  IREE--LLVMGPU/CUDA, LLVMGPU/HIP, SPIR-V/Vulkan, SPIR-V/Metal.
+  For the _same_ GPU, we typically have two paths to target, e.g., CUDA/HIP
+  or Vulkan for NVIDIA/AMD GPUs, Metal or Vulkan for Apple GPUs.
+  If one path is correct/performant, we can diff against it to try isolate
+  the problem--the common/shared compiler/runtime code is likely okay; what
+  differs between paths is likely problematic.
+* `[correctness/performance][vulkan]` Vulkan supports different GPUs.
+  Similarly, if one GPU gives correct/performant result, we diff against it
+  to find clues.
+  Even more code in compiler/runtime are shared here; what's problematic is
+  likely different capabilities triggering different CodeGen pipelines so
+  revealing bugs in a particular CodeGen pipeline.
+  Or there are driver issues from a particular vendor.
+* `[correctness]` If the CPU is working properly, we can use the same dispatch
+  region formation and diff against the CPU dispatches one by one to isolate
+  the problem.
```

Comment:
Do we have an example of this that we could link to? I wouldn't block this PR on writing something new, but for some of these maybe we could even link to existing github issues or discord discussion threads where developers have gone through debugging/troubleshooting steps?

Reply:
Added a few examples. I'll try to add more examples later.